src/mapping/recurrences/attractor_mapping_recurrences.jl

```diff
@@ -0,0 +1,326 @@
+    ps = generic_mapper_print(io, mapper)
+    println(io, rpad(" system: ", ps), nameof(typeof(mapper.ds)))
+    println(io, rpad(" grid: ", ps), mapper.grid)
+    println(io, rpad(" attractors: ", ps), mapper.bsn_nfo.attractors)
+    return
+end
+
+extract_attractors(m::AttractorsViaRecurrences) = m.bsn_nfo.attractors
+
+"""
+    basins_of_attraction(mapper::AttractorsViaRecurrences; show_progress = true)
+
+This is a special method of `basins_of_attraction` that using recurrences does
+_exactly_ what is described in the paper by Datseris & Wagemakers [Datseris2022](@ref).
+By enforcing that the internal grid of `mapper` is the same as the grid of initial
+conditions to map to attractors, the method can further utilize found exit and attraction
+basins, making the computation faster as the grid is processed more and more.
+"""
+function basins_of_attraction(mapper::AttractorsViaRecurrences; show_progress = true)
+    basins = mapper.bsn_nfo.basins
+    if basins isa SparseArray;
+        throw(ArgumentError("""
+            Sparse version of AttractorsViaRecurrences is incompatible with
+            `basins_of_attraction(mapper)`."""
+        ))
+    end
+
+    if (mapper.bsn_nfo.grid_nfo isa SubdivisionBasedGrid)
+        grid = mapper.grid.max_grid
+    else
+        grid = mapper.grid.grid
+    end
+
+    I = CartesianIndices(basins)
+    progress = ProgressMeter.Progress(
+        length(basins); desc = "Basins of attraction: ", dt = 1.0
+    )
+
+    # TODO: Here we can have a slightly more efficient iteration by
+    # iterating over `I` in different ways. In this way it always starts from the edge of
+    # the grid, which is the least likely location for attractors. We need to
+    # iterate I either randomly or from its center.
+    for (k, ind) in enumerate(I)
+        if basins[ind] == 0
+            show_progress && ProgressMeter.update!(progress, k)
+            y0 = generate_ic_on_grid(grid, ind)
+            basins[ind] = recurrences_map_to_label!(
+                mapper.bsn_nfo, mapper.ds, y0; show_progress, mapper.kwargs...
+            )
+        end
+    end
+
+    # remove attractors and rescale from 1 to max number of attractors
+    ind = iseven.(basins)
+    basins[ind] .+= 1
+    basins .= (basins .- 1) .÷ 2
+    return basins, mapper.bsn_nfo.attractors
+end
+
+#####################################################################################
+# Definition of `BasinInfo` and initialization
+#####################################################################################
+abstract type Grid end
+
+mutable struct BasinsInfo{D, G<:Grid, Δ, T, A <: AbstractArray{Int, D}}
+    basins::A # sparse or dense
+    grid_nfo::G
+    Δt::Δ
+    state::Symbol
+    current_att_label::Int
+    visited_cell_label::Int
+    consecutive_match::Int
+    consecutive_lost::Int
+    prev_label::Int
+    safety_counter::Int
+    attractors::Dict{Int, StateSpaceSet{D, T}}
+    visited_list::Vector{CartesianIndex{D}}
```

Comment:
`visited_list` can be renamed as `visited_cells` or `visited_cells_list`. Technically it is not a list, it is a stack.